mern-ecommerce/controllers/auth.js

```diff
@@ -80,3 +80,81 @@ exports.isAdmin = (req, res, next) => {
   }
   next();
 };
+
+exports.checkUser = (req, res, next) => {
+  let user = req.profile && req.auth && req.profile._id == req.auth._id;
+  if (!user) {
+    return res.status(403).jssosn({
+      error: 'Access denied',
+    });
+  }
+  next();
+};
+
+exports.forgotPassword = (req, res) => {
+  if (!req.body) return res.status(400).json({ message: 'No request body' });
+  if (!req.body.email)
+    return res.status(400).json({ message: 'No Email in request body' });
+
+  console.log('forgot password finding user with that email');
+  const { email } = req.body;
+  console.log('signin req.body', email);
+
+  User.findOne({ email }, (err, user) => {
+    if (err || !user)
+      return res.status('401').json({
+        error: 'User with that email does not exist!',
+      });
+
+    const token = jwt.sign(
+      { _id: user._id, iss: 'NODEAPI' },
+      process.env.JWT_SECRET
+    );
+
+    const emailData = {
+      from: '',
+      to: email,
+      subject: 'Password Reset Instructions',
+      text: `Please use the following link to reset your password: ${
+        process.env.CLIENT_URL
+      }/reset-password/${token}`,
+      html: `<p>Please use the following link to reset your password:</p> <p>${
+        process.env.CLIENT_URL
+      }/reset-password/${token}</p>`,
+    };
+
+    return user.update
+
+  });
+};
+
+exports.resetPassword = (req, res) => {
+  const { resetPasswordLink, newPassword } = req.body;
+
+  User.findOne({ resetPasswordLink }, (err, user) => {
+    if (err || !user)
+      return res.status('401').json({
+        error: 'Invalid Link!',
+      });
+
+    const updatedFields = {
+      password: newPassword,
+      resetPasswordLink: '',
+    };
+
+    user = _.extend(user, updatedFields);
+    user.updated = Date.now();
+
+    user.save((err, result) => {
+      if (err) {
+        return res.status(400).json({
+          error: errorHandler(err),
```

Comment:
The variable `err` is not checked for null or undefined before passing it to `errorHandler`. This could lead to a TypeError if `err` is null or undefined. Consider adding a null check before calling `errorHandler`.